python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,474 @@ def array_to_vector(col: Column) -> Column:
+        signature = inspect.signature(predict_fn)
+        num_expected = len(signature.parameters)
+
+        # convert sparse input_tensor_shapes to dense if needed
+        input_shapes: list[list[int] | None]
+        if isinstance(input_tensor_shapes, Mapping):
+            input_shapes = [None] * num_expected
+            for index, shape in input_tensor_shapes.items():
+                input_shapes[index] = shape
+        else:
+            input_shapes = input_tensor_shapes  # type: ignore
+
+        # iterate over partition, invoking predict_fn with ndarrays
+        for partition in data:
+            has_tuple = isinstance(partition, Tuple)  # type: ignore
+            has_tensors = _has_tensor_cols(partition)
+
+            # require input_tensor_shapes for any tensor columns
+            if has_tensors and not input_shapes:
+                raise ValueError("Tensor columns require input_tensor_shapes")
+
+            for batch in _batched(partition, batch_size):
+                num_actual = len(batch.columns)
+                if num_actual == num_expected and num_expected > 1:
+                    # input column per expected input, convert each column into param
+                    multi_inputs = [batch[col].to_numpy() for col in batch.columns]
+                    if input_shapes:
+                        if len(input_shapes) == num_actual:
+                            multi_inputs = [
+                                np.vstack(v).reshape([-1] + input_shapes[i])  # type: ignore
+                                if input_shapes[i]
+                                else v
+                                for i, v in enumerate(multi_inputs)
+                            ]
+                        else:
+                            raise ValueError("input_tensor_shapes must match columns")
+
+                    # run model prediction function on transformed (numpy) inputs
+                    preds = predict_fn(*multi_inputs)
+                elif num_expected == 1:
+                    # multiple input columns for single expected input
+                    if has_tensors:
+                        # tensor columns
+                        if len(batch.columns) == 1:
+                            # one tensor column and one expected input, vstack rows
+                            single_input = np.vstack(batch.iloc[:, 0])
+                        else:
+                            raise ValueError(
+                                "Multiple input columns found, but model expected a single "
+                                "input, use `struct` or `array` to combine columns into tensors."
+                            )
+                    else:
+                        # scalar columns
+                        if len(batch.columns) == 1:
+                            # single scalar column, remove extra dim
+                            single_input = np.squeeze(batch.to_numpy())
```

Comment:
For this case, I think we should force the input_shapes[0] to be None or [] ?

Reply:
I'm not sure we need this, since the input_shapes are intended to be a hint from the user to the code, so if the user's hint is incorrect, it should be fixed.  Also, in this specific case, the column is expected to be a scalar, so the user should not supply input_tensor_shapes.

Reply:
> the input_shapes are intended to be a hint from the user to the code, so if the user's hint is incorrect, it should be fixed.

I agree this. But my concern is if user accidentally set wrong input shape and code run into this branch, it will cause error which is hard to understand, so I prefer to add strict verification and fail-fast. E.g., in this case,
we need to verify input_shapes[0] must be None (user not set) or `[]` (means scalar shape), WDYT ?

Reply:
Sounds reasonable, added some more more validations to cover this.